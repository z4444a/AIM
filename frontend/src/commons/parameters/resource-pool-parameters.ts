import ResourcePoolModel from '../../model/get/resource-pool-model';
import { PageAndSortParams } from '../../model/parameters/page-and-sort-params';

export default interface ResourcePoolParameters extends PageAndSortParams<ResourcePoolModel> {
  name?: string;
  type?: number;
  active?: boolean | undefined;
  onlyMine?: boolean | undefined;
}
