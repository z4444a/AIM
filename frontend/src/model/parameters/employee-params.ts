import { EmployeeFullModel } from '../get/employee-full-model';
import { PageAndSortParams } from './page-and-sort-params';

export interface EmployeeParams extends PageAndSortParams<EmployeeFullModel> {
  name?: string;
}
