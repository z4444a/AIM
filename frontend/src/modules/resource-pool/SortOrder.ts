import { Order } from './Order';

export type SortOrder = Order.ASC | Order.DESC | undefined;
