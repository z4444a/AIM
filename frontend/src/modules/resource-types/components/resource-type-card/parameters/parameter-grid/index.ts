import { withTranslation } from 'react-i18next';
import { ParameterGrid } from './parameter-grid';

export default withTranslation('common')(ParameterGrid);
