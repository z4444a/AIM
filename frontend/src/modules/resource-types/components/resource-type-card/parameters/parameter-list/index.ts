import { ParameterList } from './parameter-list';
import { withTranslation } from 'react-i18next';
import withStyles from '@material-ui/core/styles/withStyles';
import { styles } from './styles';

export default withTranslation('common')(withStyles(styles)(ParameterList));
