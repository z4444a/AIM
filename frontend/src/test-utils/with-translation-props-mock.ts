import { WithTranslation } from 'react-i18next';

export function getMockWithTranslationProps(): WithTranslation {
  // eslint-disable-next-line
  const t = ((key: string) => `${key}_i18n`) as any;
  // eslint-disable-next-line
  const i18n = {} as any;
  const props: WithTranslation = {
    tReady: true,
    t,
    i18n,
  };

  return props;
}
