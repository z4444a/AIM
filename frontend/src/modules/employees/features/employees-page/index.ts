import { fetchFullEmployees } from '../../../../redux/actions/full-employee-action';
import { push } from 'connected-react-router';
import { SynchronizeAction } from '../../../../redux/actions/synchronize-action';
import { RootState } from 'typesafe-actions';
import { connect } from 'react-redux';
import { EmployeesPage } from './employees-page';
import { withTranslation } from 'react-i18next';
import { toggleEmployeePageFilterPanel } from '../../../../redux/actions/portal-actions';
import { EmployeeParams } from '../../../../model/parameters/employee-params';

export const mapStateToProps = (state: RootState) => ({
  page: state.fullEmployees.page,
  params: state.fullEmployees.params,
  showFilterPanel: state.portal.employeePageFilterPanelOpen,
});

export const mapDispatchToProps = (dispatch: Function) => ({
  fetch: (params?: EmployeeParams) => dispatch(fetchFullEmployees.request(params)),
  redirect: (path: string) => dispatch(push(path)),
  sync: (params?: EmployeeParams) => dispatch(SynchronizeAction.request(params)),
  toggleFilterPanel: (showFilterPanel: boolean) =>
    dispatch(toggleEmployeePageFilterPanel(showFilterPanel)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation('common')(EmployeesPage));
