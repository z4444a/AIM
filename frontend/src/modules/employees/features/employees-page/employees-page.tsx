import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import React from 'react';
import { Path } from '../../../../commons/path';
import { EmployeeFullModel } from '../../../../model/get/employee-full-model';
import Sync from '@material-ui/icons/Sync';
import {
  FilterValueType,
  Props as DataGridProps,
} from '../../../../commons/components/aim-data-grid/aim-data-grid';
import AimDataGrid from '../../../../commons/components/aim-data-grid';
import { SortOrder } from '../../../resource-pool/SortOrder';
import { MUIDataTableColumn } from 'mui-datatables';
import { WithTranslation } from 'react-i18next';
import { mapDispatchToProps, mapStateToProps } from './index';
import EmployeesFilterPanel from '../../components/employees-filter-panel/index';
import { FilterField } from '../../components/employees-filter-panel/employees-filter-panel';
import { Role } from '../../../../commons/role';
import { EmployeeParams } from '../../../../model/parameters/employee-params';

export type Props = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps> &
  WithTranslation;

export class EmployeesPage extends React.Component<Props> {
  private readonly colDefs: MUIDataTableColumn[];
  private readonly gridConfig: Partial<DataGridProps>;

  public constructor(props: Props) {
    super(props);
    this.gridConfig = this.getGridConfig();
    this.colDefs = this.getGridColumnsConfig();
  }

  public componentDidMount(): void {
    this.props.fetch(this.props.params);
  }

  public render(): JSX.Element {
    const { page, params, showFilterPanel } = this.props;
    const sort = {
      sortBy: params.sortBy as string,
      direction: params.order,
    };

    if (!page) {
      return <div />;
    }

    const filter: { [key: string]: FilterValueType } = {};
    if (params.name) {
      filter[FilterField.EMPLOYEE_LAST_NAME] = params.name;
    }
    return (
      <AimDataGrid
        columns={this.colDefs}
        data={page.content.map(this.mapEmployeeRoles)}
        totalPageCount={page.totalElements}
        rowsPerPage={params.pageSize}
        sort={sort}
        filterValues={filter}
        filterPanelOpen={showFilterPanel}
        {...this.gridConfig}
      />
    );
  }

  private getGridConfig(): Partial<DataGridProps> {
    const { t } = this.props;
    const gridConfig: Partial<DataGridProps> = {
      pageTitle: t('employeePage.header'),
      onColumnSortChange: (changedColumn, direction) =>
        this.handleColumnSortChange(changedColumn, direction),
      onChangePage: currentPage => this.handleChangePage(currentPage),
      onChangeRowsPerPage: numberOfRows => this.handleChangeRowsPerPage(numberOfRows),
      onRowClick: (rowData, rowMeta, dataItem) => this.handlerRowClick(rowData, rowMeta, dataItem),
      renderCustomFilterCmp: context => <EmployeesFilterPanel context={context} />,
      onFilterChange: this.handleFilterChange,
      toggleFilterPanel: this.toggleShowFilterPanel,
      customToolbar: () => {
        return (
          <Tooltip title={t('employeePage.buttons.syncToIntranet')}>
            <IconButton onClick={this.handleSync}>
              <Sync />
            </IconButton>
          </Tooltip>
        );
      },
    };

    return gridConfig;
  }

  private toggleShowFilterPanel = (showFilterPanel: boolean) => {
    const { toggleFilterPanel } = this.props;
    toggleFilterPanel(showFilterPanel);
  };

  private getGridColumnsConfig(): MUIDataTableColumn[] {
    const { t } = this.props;
    const colDefs: MUIDataTableColumn[] = [
      {
        label: t('employeePage.grid.columns.id'),
        name: 'id',
      },
      {
        label: t('employeePage.grid.columns.lastName'),
        name: 'lastName',
      },
      {
        label: t('employeePage.grid.columns.firstName'),
        name: 'firstName',
      },
      {
        label: t('employeePage.grid.columns.middleName'),
        name: 'middleName',
      },
      {
        label: t('employeePage.grid.columns.position'),
        name: 'post',
      },
      {
        label: t('employeePage.grid.columns.role'),
        name: 'role',
      },
    ];
    return colDefs;
  }

  private handleColumnSortChange(changedColumn: string, direction: SortOrder) {
    const changeParams: Partial<EmployeeParams> = {
      sortBy: changedColumn as keyof EmployeeFullModel,
      order: direction,
    };
    this.handleChange(changeParams);
  }

  private handleChangePage(currentPage: number) {
    const changeParams: Partial<EmployeeParams> = {
      page: currentPage,
    };
    this.handleChange(changeParams);
  }

  private handleChangeRowsPerPage(numberOfRows: number) {
    const changeParams: Partial<EmployeeParams> = {
      pageSize: numberOfRows,
    };
    this.handleChange(changeParams);
  }

  private handlerRowClick(
    rowData: string[],
    rowMeta: { dataIndex: number; rowIndex: number },
    dataItem: object
  ) {
    this.handleRedirect(dataItem as EmployeeFullModel);
  }

  private handleChange(changeParams?: Partial<EmployeeParams>): void {
    const params: EmployeeParams = Object.assign({}, this.props.params, changeParams);
    this.props.fetch(params);
  }

  private handleFilterChange = (values: { [key: string]: FilterValueType }) => {
    const changeParams: Partial<EmployeeParams> = {
      name: values[FilterField.EMPLOYEE_LAST_NAME]
        ? values[FilterField.EMPLOYEE_LAST_NAME] + ''
        : undefined,
    };
    this.handleChange(changeParams);
  };

  private handleRedirect(item: EmployeeFullModel) {
    const id: string = item.id.toString(10);
    this.props.redirect(Path.EMPLOYEE_CARD.replace(':id', id));
  }

  private handleSync = () => {
    this.props.sync(this.props.params);
  };

  private mapEmployeeRoles = (employee: EmployeeFullModel) => {
    const { t } = this.props;
    const role =
      [Role.POOL_CREATOR, Role.POOL_OWNER].indexOf(employee.role) >= 0 ? Role.USER : employee.role;
    return {
      ...employee,
      role: t('employeeChangeRolePage.' + role),
    };
  };
}
