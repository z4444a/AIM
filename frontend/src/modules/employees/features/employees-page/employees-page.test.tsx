import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { getMockWithTranslationProps } from '../../../../test-utils/with-translation-props-mock';
import { EmployeesPage, Props } from './employees-page';
import { Order } from '../../../resource-pool/Order';
import { EmployeeParams } from '../../../../model/parameters/employee-params';

describe('EmployeesPage', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const withTranslationProps = getMockWithTranslationProps();
    const params: EmployeeParams = {
      order: Order.ASC,
      pageSize: 10,
      sortBy: 'firstName',
      page: 0,
      name: '',
    };
    const page = {
      content: [],
      totalElements: 10,
      page: 0,
    };
    const props: Props = {
      params,
      redirect: () => null,
      fetch: () => null,
      sync: () => null,
      toggleFilterPanel: () => null,
      page,
      ...withTranslationProps,
      showFilterPanel: true,
    };
    wrapper = shallow(<EmployeesPage {...props} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
