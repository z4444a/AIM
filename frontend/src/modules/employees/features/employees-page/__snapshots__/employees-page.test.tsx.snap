// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeesPage should render correctly 1`] = `
<withI18nextTranslation(WithStyles(AimDataGrid))
  columns={
    Array [
      Object {
        "label": "employeePage.grid.columns.id_i18n",
        "name": "id",
      },
      Object {
        "label": "employeePage.grid.columns.lastName_i18n",
        "name": "lastName",
      },
      Object {
        "label": "employeePage.grid.columns.firstName_i18n",
        "name": "firstName",
      },
      Object {
        "label": "employeePage.grid.columns.middleName_i18n",
        "name": "middleName",
      },
      Object {
        "label": "employeePage.grid.columns.position_i18n",
        "name": "post",
      },
      Object {
        "label": "employeePage.grid.columns.role_i18n",
        "name": "role",
      },
    ]
  }
  customToolbar={[Function]}
  data={Array []}
  filterPanelOpen={true}
  filterValues={Object {}}
  onChangePage={[Function]}
  onChangeRowsPerPage={[Function]}
  onColumnSortChange={[Function]}
  onFilterChange={[Function]}
  onRowClick={[Function]}
  pageTitle="employeePage.header_i18n"
  renderCustomFilterCmp={[Function]}
  rowsPerPage={10}
  sort={
    Object {
      "direction": "asc",
      "sortBy": "firstName",
    }
  }
  toggleFilterPanel={[Function]}
  totalPageCount={10}
/>
`;
